### DISCUSSION TITLE
Property hooks with attribute

### DISCUSSION THREAD

**Discussion by Cologler** [canton7/PropertyChanged.SourceGenerator#5] [Ideas]
Currently, property hook work like:

``` cs
[Notify] private string _firstName;
[Notify] private string _lastName;
[Notify] private string _fullName;

void OnFirstNameChanged() => this.UpdateFullName();
void OnLastNameChanged() => this.UpdateFullName();
void UpdateFullName() => ...
```

If the `FullName` depends on more properties, the `On{PropertyName}Changed()` list may grow unnecessary.

And the property hook is hard to detect errors in compile-time if we try to rename the `_firstName`.

What if we can use attributes, it can be:

``` cs
[Notify][PropertyChangedHook(nameof(UpdateFullName))] private string _firstName;
[Notify][PropertyChangedHook(nameof(UpdateFullName))] private string _lastName;
[Notify]private string _fullName;

void UpdateFullName() => ...
```

**Comment by Cologler** [comment]
@canton7

**Reply by canton7** [reply]
Sorry, real life is taking a bunch of my time.

Possible -- it wouldn't be *that* hard to add, but I might question the need.

In this specific case, wouldn't it make more sense to write it as:

```cs
[Notify] private string _firstName;
[Notify] private string _lastName;
public string FullName => $"{_firstName} {_lastName}";
```

That uses the automatic property dependency stuff, and a `PropertyChanged` event for `FullName` will be raised whenever `FirstName` or `LastName` changes. If you need to use a method to generate the value for `FullName`, you can do:

```cs
[Notify] private string _firstName;
[Notify] private string _lastName;
[DependsOn(nameof(FullName), nameof(LastName))] public string FullName => CreateFullName();
```

(or the equivalent using `AlsoNotify`).

These both have the advantage of letting `FullName` say what it depends on, rather than forcing you to manually keep track of this.

The downside is that you don't cache the value of `FullName` and it's re-calculated on each access, but I'd love to see a case where that actually mattered!

**Reply by Cologler** [reply]
Don't need to apologize, real life is more important.

Unlink the `FullName` is a property, here is a partial code of my real-world project:

``` cs
        private void UpdateExpirationsEnd()
        {
            ...
        }

        private void OnExpirationsStartChanged() => this.UpdateExpirationsEnd();

        private void OnExpirationsSelectedTimeSpanChanged() => this.UpdateExpirationsEnd();

        private void OnExpirationsTimeSpanUnitSelectedItemChanged() => this.UpdateExpirationsEnd();
```

Since the `UpdateExpirationsEnd()` is not a property, it seems to be unable to use the `[AlsoNotify]` attribute.

**Reply by canton7** [reply]
Fair enough. I can't promise when I'll get time to look at this, it might be a few weeks.

If you fancy having a crack yourself, [`OnPropertyNameChanged`](https://github.com/canton7/PropertyChanged.SourceGenerator/blob/cb9deb2f0ca15984f3e34f147c9c9e525ba1b5fa/src/PropertyChanged.SourceGenerator/Analysis/MemberAnalysis.cs#L16) will need to change to a `List<OnPropertyNameChangedInfo>`, and [`FindPropertyNameChangedMethod`](https://github.com/canton7/PropertyChanged.SourceGenerator/blob/cb9deb2f0ca15984f3e34f147c9c9e525ba1b5fa/src/PropertyChanged.SourceGenerator/Analysis/Analyser_RaiseMethods.cs#L142) updated to find multiple methods to call, using a new attribute.

For the attribute name, I'm not sure whether to go with your `PropertyChangedHook` (which does match the wording in the README), or something like `OnChanged`...

**Reply by Cologler** [reply]
How about `[AlsoCall]` or allow `[DependsOn]` to add to the method?
